src/cis2/types.rs

```diff
@@ -0,0 +1,645 @@
+    fn div_assign(&mut self, other: Self) { self.0 /= other.0 }
+}
+
+impl Serial for TokenAmount {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        let mut value = self.0.clone();
+        for _ in 0..37 {
+            let mut byte = (value.clone() % BigUint::from(128u8)).to_u8().unwrap(); // Safe to unwrap since we have truncated
+            value >>= 7;
+            if !value.is_zero() {
+                byte |= 0b1000_0000;
+            }
+            out.write_u8(byte)?;
+
+            if value.is_zero() {
+                return Ok(());
+            }
+        }
+        Err(W::Err::default())
+    }
+}
+
+impl Deserial for TokenAmount {
+    fn deserial<R: Read>(source: &mut R) -> concordium_contracts_common::ParseResult<Self> {
+        let mut result = BigUint::zero();
+        for i in 0..37 {
+            let byte = source.read_u8()?;
+            let value_byte = BigUint::from(byte & 0b0111_1111);
+            result += value_byte << (i * 7);
+
+            if byte & 0b1000_0000 == 0 {
+                return Ok(TokenAmount::from(result));
+            }
+        }
+        Err(concordium_contracts_common::ParseError {})
+    }
+}
+
+/// CIS2 Token ID can be up to 255 bytes in size.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct TokenIdVec(pub Vec<u8>);
+
+/// Error from parsing a token ID bytes from a hex encoded string.
+#[derive(Debug, Error)]
+pub enum ParseTokenIdVecError {
+    #[error("Invalid hex string: {0}")]
+    ParseIntError(#[from] hex::FromHexError),
+    #[error("Token ID too large. Maximum allowed size is 255 bytes. {0} bytes were provided.")]
+    TooManyBytes(usize),
+}
+
+/// Parse a Token ID from a hex encoded string.
+impl FromStr for TokenIdVec {
+    type Err = ParseTokenIdVecError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let bytes = hex::decode(s)?;
+        if bytes.len() > 255 {
+            Err(ParseTokenIdVecError::TooManyBytes(bytes.len()))
+        } else {
+            Ok(TokenIdVec(bytes))
+        }
+    }
+}
+
+/// Display the token ID as a hex string.
+impl std::fmt::Display for TokenIdVec {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        for b in &self.0 {
+            write!(f, "{:02x}", b)?;
+        }
+        Ok(())
+    }
+}
+
+/// Serialize the token ID according to CIS2 specification.
+impl Serial for TokenIdVec {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        let len = u8::try_from(self.0.len()).map_err(|_| W::Err::default())?;
+        len.serial(out)?;
+        serial_vector_no_length(&self.0, out)
+    }
+}
+
+/// Deserialize bytes to a Token ID according to CIS2 specification.
+impl Deserial for TokenIdVec {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let tokens_id_length = u8::deserial(source)?;
+        let bytes = deserial_vector_no_length(source, tokens_id_length.into())?;
```

Comment:
Why are you not checking length here?

Reply:
I am not sure what to check for, since any value of the u8 is a valid length for the token id

Reply:
My mistake.